command/monitor/monitor.go

```diff
@@ -92,32 +92,28 @@ func (d *monitor) Start() <-chan []byte {
 		ticker := time.NewTicker(d.dropCheckInterval)
 		defer ticker.Stop()
 
+		var logMessage []byte
 		for {
-			// This will initialize logMessage to its zero value (empty slice)
-			// on every loop iteration, which is what makes the below for loop
-			// work.
-			var logMessage []byte
-
-			for len(logMessage) == 0 {
-				select {
-				case <-ticker.C:
-					// Check if there have been any dropped messages.
-					dc := atomic.LoadUint64(&d.droppedCount)
-
-					if dc > 0 {
-						logMessage = []byte(fmt.Sprintf("[WARN] Monitor dropped %d logs during monitor request\n", dc))
-						atomic.SwapUint64(&d.droppedCount, 0)
-					}
-				case logMessage = <-d.logCh:
-				case <-d.doneCh:
-					return
-				}
-			}
+			// Reset the byte slice on every loop iteration, which is what makes
+			// the below for loop work.
+			logMessage = nil
 
 			select {
+			case <-ticker.C:
+				// Check if there have been any dropped messages.
+				dc := atomic.LoadUint64(&d.droppedCount)
+
+				if dc > 0 {
+					logMessage = []byte(fmt.Sprintf("[WARN] Monitor dropped %d logs during monitor request\n", dc))
+					atomic.SwapUint64(&d.droppedCount, 0)
+				}
+			case logMessage = <-d.logCh:
 			case <-d.doneCh:
 				return
-			case streamCh <- logMessage:
+			}
+
+			if len(logMessage) > 0 {
```

Comment:
We still need a second select so that we can ensure that we return when doneCh is closed instead of being blocked on streamCh.

Reply:
Good catch

Reply:
Actually, if we were to be blocked by the `streamCh`, it would have already fallen into the case so the `doneCh` case wouldn't unblock it.

Reply:
Nevermind, tested this and that's not the case.